File: app/src/main/java/com/hvasoft/pokedex/presentation/ui/common/UIExtensionFunctions.java
package com.hvasoft.pokedex.presentation.ui.common;

import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.google.android.material.snackbar.Snackbar;
import com.hvasoft.pokedex.R;

public class UIExtensionFunctions {

    public static void showPopUpMessage(View view, String message, Boolean isError) {
        int duration = isError ? Snackbar.LENGTH_LONG : Snackbar.LENGTH_SHORT;

        Snackbar snackbar = Snackbar.make(view, message, duration);
        ViewGroup.LayoutParams params = snackbar.getView().getLayoutParams();
        if (params instanceof ViewGroup.MarginLayoutParams marginParams) {
            int extraBottomMargin =
                    view.getContext().getResources().getDimensionPixelSize(R.dimen.common_padding_default);
            marginParams.setMargins(
                    marginParams.leftMargin,
                    marginParams.topMargin,
                    marginParams.rightMargin,
                    marginParams.bottomMargin + extraBottomMargin
            );
            snackbar.getView().setLayoutParams(marginParams);
        }

        snackbar.show();
    }

    public static void loadImageWithUrl(ImageView imageView, String imageUrl) {
        Glide.with(imageView.getContext())
                .load(imageUrl)
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .centerInside()
                .placeholder(R.drawable.ic_placeholder)
                .error(R.drawable.ic_caution)
                .into(imageView);
    }

}

File: app/src/main/java/com/hvasoft/pokedex/presentation/ui/common/UIExtensionFunctions.kt
package com.hvasoft.pokedex.presentation.ui.common

import android.graphics.drawable.Drawable
import android.view.ViewGroup
import android.widget.ImageView
import androidx.fragment.app.Fragment
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.target.CustomTarget
import com.bumptech.glide.request.transition.Transition
import com.google.android.material.snackbar.Snackbar
import com.hvasoft.pokedex.R

fun Fragment.showPopUpMessage(anyMessage: Any, isError: Boolean = false) {
    val duration = if (isError) Snackbar.LENGTH_LONG else Snackbar.LENGTH_SHORT
    val message = if (anyMessage is Int) getString(anyMessage) else anyMessage.toString()
    view?.let { rootView ->
        val snackBar = Snackbar.make(rootView, message, duration)
        val params = snackBar.view.layoutParams as ViewGroup.MarginLayoutParams
        val extraBottomMargin = resources.getDimensionPixelSize(R.dimen.dimen_64)
        params.setMargins(
            params.leftMargin,
            params.topMargin,
            params.rightMargin,
            params.bottomMargin + extraBottomMargin
        )
        snackBar.view.layoutParams = params
        snackBar.show()
    }
}

fun ImageView.loadImageWithUrl(
    url: String?,
    isCircle: Boolean = false,
    onSuccess: () -> Unit = {},
    onError: () -> Unit = {}
) {
    var glideRequest = Glide.with(this)
        .load(url)
        .diskCacheStrategy(DiskCacheStrategy.ALL)
        .centerCrop()
    if (isCircle) glideRequest = glideRequest.circleCrop()
    glideRequest
        .into(object : CustomTarget<Drawable>() {
            override fun onResourceReady(resource: Drawable, transition: Transition<in Drawable>?) {
                setImageDrawable(resource)
                onSuccess()
            }

            override fun onLoadCleared(placeholder: Drawable?) {
                onError()
            }

            override fun onLoadFailed(errorDrawable: Drawable?) {
                onError()
            }
        })
}

fun String.isInputInitialValid(): Boolean {
    if (isEmpty())
        return false
    val words = split(" ")
    if (words.isNotEmpty()) {
        val firstWord = words.first()
        if (firstWord.isNotEmpty() && firstWord.first().isLetterOrDigit())
            return true
    }
    return false
}

fun String.getInitials(): String {
    val words = split(" ")
    if (words.isNotEmpty()) {
        val firstWord = words.first()
        if (firstWord.isNotEmpty() && firstWord.first().isLetterOrDigit())
            return firstWord.first().toString()
    }
    return ""
}